avalanche/training/strategies/icarl.py

```diff
@@ -187,19 +189,21 @@ def compute_class_means(self, strategy):
 
     def construct_exemplar_set(self, strategy: BaseStrategy):
         tid = strategy.clock.train_exp_counter
-        nb_cl = strategy.experience.benchmark.n_classes_per_exp
+        benchmark = strategy.experience.benchmark
+        nb_cl = benchmark.n_classes_per_exp[tid]
+        previous_seen_classes = sum(benchmark.n_classes_per_exp[:tid])
 
         if self.fixed_memory:
             nb_protos_cl = int(ceil(
                 self.memory_size / len(self.observed_classes)))
         else:
             nb_protos_cl = self.memory_size
-        new_classes = self.observed_classes[tid * nb_cl[tid]:
-                                            (tid + 1) * nb_cl[tid]]
+        new_classes = self.observed_classes[previous_seen_classes:
+                                            previous_seen_classes + nb_cl]
```

Comment:
new classes may not be contiguous in Avalanche

Reply:
I think class labels should be treated as sets since Avalanche benchmarks may have unordered labels.

Reply:
What do you mean by "new classes may not be contiguous in avalanche"? The iCaRL plugin in avalanche build the `self.observed_classes` field by inserting the classes from `benchmark.class_order`. If you mean that this does not work in a scenario which is not class incremental I agree, but we should reformat all the strategy, since by now iCaRL only works with class-incremental plugins.

Reply:
Does the strategy works if you have Exp.1 with classes [2, 3], and Exp.2 with classes [0, 1]? If so, you can ignore my comment.

Reply:
The behavior of the strategy is the same as before, I just "generalize" the behavior for working with benchmarks with different number of classes per experience.

Reply:
> Does the strategy works if you have Exp.1 with classes [2, 3], and Exp.2 with classes [0, 1]? If so, you can ignore my comment.

I guess, yes, as long as the benchmark has the class order set to [2,3,1,0]